### ISSUE TITLE
Limit number of edits per user and day

### ISSUE THREAD

**Issue by woodpeck** [openstreetmap/openstreetmap-website#2342]
It regularly happens that people - usually, but not always new signups - upload hundreds of thousands of objects to OSM before someone notices and tells them to stop. Then we have to delete those hundreds of thousands of objects again. (Case in point from recent past, https://www.openstreetmap.org/user/maxiangying.) This is undesirable:
* it skews statistics ("wow, today was the day with the most edits since OSM started!!!" etc)
* it wastes processing time and bandwidth on the thousands of servers world wide that are configured to consume updates from OSM
* it wastes processing time and bandwidth on our own database server (where objects are first created, then deleted again, and a copy of the object is nonetheless held in history)
* it wastes the time of volunteers who have to remove the edits
* it embarrasses the person who uploaded the data (at least I hope it does)

While editors can, and should, inform their users about potential issues, I think it would also be worth contemplating to have some sort of rate limit on the API. It could be something that users can override but not accidentally - for example, you could be normally limited to X edits per day (exact numbers t.b.d.) and then you could click a button in your user preferences that says "I have read the data import and mechanical edit guidelines and I want to lift the limitation for one week" or so.

**Comment by tomhughes** [comment]
The production API is in the cgimap repository, not here.

**Comment by woodpeck** [comment]
Closed in favour of https://github.com/zerebubuth/openstreetmap-cgimap/issues/207

**Comment by tomhughes** [comment]
From a practical point of view you need to define "edit" somehow - the obvious answer is changesets but that can be easily gamed so it will presumably need to be the number of objects added/changed/deleted.

Doing it per day is probably excessively expensive from an implementation standpoint because it requires resetting counters at some point, or keeping say hourly counts and expiring those over a day old. A token bucket type algorithm like the download limit probably makes more sense.

From a personal point of view I dread dealing with the resulting complaints...

**Comment by simonpoole** [comment]
Actually I think the issue should remain here :-). 

At least the obvious, how stupid me would do it, implementation would involve automatically creating a block once the limit has been exceeded (because all the bits and pieces are already there and we don't want to create a slightly different implementation of the same thing).

I believe this would address @tomhughes  concerns to as the messaging can be clear what to do to be unblocked (OK I believe we still have the issue with iD borking when an account has been blocked but iD users are unlikely to run in to this anyway).

**Comment by tomhughes** [comment]
It doesn't matter how clear you make the messaging, you will get people pleading for themselves to be special cased because, you know, PEOPLE MIGHT DIE.

**Comment by mmd-osm** [comment]
Just to add a bit of detail here, we already have rate limiting on the changeset upload call on cgimap, which is currently based on the number of bytes returnes by the call (like it is done for the /map call).  Once you hit the limit, the upload would be rejected via HTTP 509 Bandwidth Limit Exceeded. It's not a permanent block either, so it wouldn't place too much of a burden on sysadmins.

Admittedly, looking at the number of bytes is not effective by any means to stop the users mentioned here (main reason being that the diffResult message usually comes with a rather tiny amount of bytes compared to a /map call). ;aybe we could evolve this approach by defining some more meaningful criteria.

**Comment by tomhughes** [comment]
What I was proposing was some sort of equivalent limit on the upload side, whether that's just bytes uploaded, or number of objects changed. But implemented in basically the same way.

**Comment by gravitystorm** [comment]
> we already have rate limiting on the changeset upload call on cgimap

Where does cgimap store the rate limit status? Is it in the database somewhere, or does each instance of cgimap (i.e. on different backend servers) keep track of the rates separately?

> The production API is in the cgimap repository, not here.

I think it's worth discussing it here too, particularly if we want to ensure clear messaging to the user when they go over any quota.

> From a personal point of view I dread dealing with the resulting complaints...

So let's make sure the complaints go to someone else! I think it would be something for DWG to deal with, rather than the sysadmin group.

**Comment by tomhughes** [comment]
They're stored in memcache, so shared across instances but can be lost if a memcache server restarts.

**Comment by mmd-osm** [comment]
> I think it's worth discussing it here too, particularly if we want to ensure clear messaging to the user when they go over any quota.

For a clear messaging we'd need something better than _You have downloaded too much data. Please try again later._ with comes with an  _HTTP 509 Bandwidth Limit Exceeded_ error response.

Also, both downloads and uploads currently share the same user-based key in memcache. A user downloading lots of data might be impacted when trying to upload some changes (and vice versa). I don't know if this is good idea. On the other hand, if memcache memory permits, we might as well introduce a dedicated key only used for changeset upload tracking.

**Comment by mmd-osm** [comment]
> It could be something that users can override but not accidentally - for example, you could be normally limited to X edits per day (exact numbers t.b.d.) and then you could click a button in your user preferences that says "I have read the data import and mechanical edit guidelines and I want to lift the limitation for one week" or so.

I still don't have a clear picture what makes sense here. Checking the users table for some timestamp when rate limiting restrictions would be back to normal, isn't impossible to implement. Pretty much the same applies to the user registration date or number of changesets as additional decision criteria.
It's mostly a matter of how a meaningful policy could look like, and also this requires some consensus building in the community.

**Comment by woodpeck** [comment]
It is certainly good to air the idea with the community a bit, though when it comes to it, I would prefer to implement a technologically viable mechanism (that doesn't make the source more complex than necessary and doesn't waste to many resources when executed) and then have the community participate in parametrising that, rather than attempt to find consensus on how exactly something should be implemented. 

We're not writing a bill of rights here; I think it will be easy to find consensus on the bare outline that I specified initially (make it so that we can't have someone sign up and upload 200k buildings in two days before anybody even notices), and anything more detailed could well be entrusted to those who actually write code ;)

**Comment by mmd-osm** [comment]
> then you could click a button in your user preferences that says "I have read the data import and mechanical edit guidelines and I want to lift the limitation for one week" or so.

You could argue that large scale imports come with some kind of responsibility, perhaps even require a commitment to the project. Instead of people clicking on a button, you might be thinking about whether this is something where an OSMF membership status could come into play.

**Comment by Zverik** [comment]
2012: your plugin for preventing people spending all their time on mapping is harmful.
2019: okay we need to limit mappers' activity.

**Comment by mmd-osm** [comment]
@Zverik : your "no more mapping" plugin was a prank to kill JOSM altogether. I don't see how this relates to this discussion. Seems rather off topic to me.

**Comment by Zverik** [comment]
It was as much of a prank as "delete the repository" button in github. And it relates to the discussion: instead of questioning why should we make arbitrary limits on uploading and why DWG cannot process or even find these super-productive users, we are discussing how to block people from uploading data to OSM automatically. First by number of changesets, then by user-agent, then by geoip and so on — temporary measures completely opaque and puzzling users, with no warnings, unlike that plugin.

**Comment by dieterdreist** [comment]
couldn’t you just create another account if you were to hit an upload limit? IP addresses are often dynamically assigned, user agents can be customized at will, VPNs are cheap, what would prevent someone from circumventing arbitrary limits anyway?

**Comment by tomhughes** [comment]
Yes obviously it's not going to stop somebody that is aware what they're doing is wrong but is determined to go ahead, but that's not the goal, because no technical solution could.

The goal is make the well intentioned but over enthusiastic people stop, think, and learn about the community rules around imports and automated edits.

**Comment by woodpeck** [comment]
> The goal is make the well intentioned but over enthusiastic people stop, think, and learn about the community rules around imports and automated edits.

Or, turning this around, today if someone uploads 200,000 buildings in one day without any prior consultation then it could have been an accident. I want to come to a point where if someone does this, it is obvious that it was done with ill intent.

**Comment by mmd-osm** [comment]
Come to think of it, we should probably focus a bit more on UX and discuss ways on how to guide (new) users without patronizing them too much.

In particular, large scale uploads, or massive moving around of nodes (based on some well meaning clean up effort) could be easily detected in JOSM and combined with other user metrics like number of changesets or day or date of registration. Also, JOSM knows how much time a user spent in an editing session and can figure out, if it's a simply Overpass download, search and replace and upload within 1 minute kind of mechanical upload, which the API would never be able to figure out.

On top, we can present fully localized messages to the user, along with a nice graphical step-by-step guide to go through some checklist or similar, rather than some technical HTTP error message the API would return.

So again, let's focus on the user experience first, think how the user interaction could look like, and what kind of guidance would be reasonable and acceptable.

**Comment by woodpeck** [comment]
Are you sure this is the right place to design the JOSM user interface? Is that not something that should be left to the JOSM developers? There's a relationship of course, but I think that we should do this like we did it with the limitation of changeset sizes: The OSM API decides what is acceptable and rejects anything unacceptable with an error message, and it is then the duty of the editors to ideally guide their users so that this error condition is never triggered, or to deal with the error in a suitable fashion should it occur.

Since I'm a JOSM user myself I could certainly sketch a number of possible JOSM UI approaches to get this right, but they would not necessarily apply to other editors that have different UIs - I think that approaching this from a UI perspective is the wrong approach for us in this repository.

We should build a mechanism that lets us reject uploads of "too much in too short time", and we could initially set these limits very high and give editor writes the chance to experiment and adapt their UI if necessary, and then tighten the limits to something sensible. I think that trying to design good UIs for all editors is out of scope here.

**Comment by mmd-osm** [comment]
I don't to go into details about the differences between UX and UI design, as it would be a bit beyond the scope of this issue. I'm 100% positive that you need to think this issue end-to-end starting from the editing app to the API to get some meaningful results, rather than just annoy users with arbitrary limits. This is mainly a JOSM issue. I clearly don't see any of the other editors anywhere close to creating a similar amount of data to upload (ignoring command line tools for a second).

From the API side, we can rather easily calculate both the number of changesets a user created during the last "n" hours, as well as as well as the total number of changes via some database statement. Number of changes would include any kind of operation (creating/updating/deleting) without knowing any further details. 

Those figures should be checked at the time when a new changeset is being created, rather than during the upload itself (I think the reasons are clear). A meaningful response should of course include the estimated waiting time, otherwise the error message won't make much sense to a user.

I would consider anything that requires any sort of domain knowledge (e.g. counting "number of houses") to be out of scope for the API. If a user uploads tons of objects with rather uncommon tags, the editor app should take care of that.

Still, the overall approach as it stands has a massive negative impact on people trying to clean up failed imports from other mappers or simply revert vandalism. I want to see a bit more discussion how those negative effects can be minimized, if not completely avoided.

**Comment by woodpeck** [comment]
I suggested in my initial message that there could be a mechanism whereby people could - perhaps temporarily - lift the limit in a self-service fashion. That would go a long way to ensure that if someone knows what they're doing they will not be hindered, but if someone does *not* know what they're doing they will be. 

I agree that all the API can do is count the number of modifications, not what exactly something is, and I never intended it to; I mentioned houses just as an example. If someone uploads a hundred thousand trees (or, also a popular option, a hundred thousand untagged objects), or *deletes* a hundred thousand houses, that's just as bad.

**Comment by mmd-osm** [comment]
(adding https://github.com/bryceco/GoMap/issues/113#issuecomment-384604159 as an example of an app creating 800 empty changesets in 3 minutes).

**Comment by richlv** [comment]
@mmd-osm, thank you for the link. This emphasises the need to have the limits in the API - even if we could get all editors to implement internal limits (plenty hard), a simple bug (perhaps in the limiting code itself...) could still cause a changeset storm.

**Comment by matkoniecz** [comment]
> Closed in favour of zerebubuth/openstreetmap-cgimap#207

@woodpeck This issue was closed, with request to reopen this one https://github.com/zerebubuth/openstreetmap-cgimap/issues/207#issuecomment-553039871

**Comment by mmd-osm** [comment]
I have outlined in the cgimap issue, why should be dealt with exclusively on the Rails side at this time: 

1. Quota check should be done when creating a new changeset, which is on Rails
2. Quota maintenance requires a UI and a database table with configuration settings -> Rails
3. UI for users to extend their quota requires a UI -> Rails

**Comment by woodpeck** [comment]
Here's a recent case where someone uploaded 3 million nodes and 300k ways in one day without knowing what they were doing, all to be reverted again: https://www.openstreetmap.org/changeset/76868869 (and older).

**Comment by Chaz6** [comment]
Here is my suggestion. Implement a process whereby 2 or more users with a minimum account age can be petitioned. Upon acceptance by the minimum quorum, a token is generated for one-time use which is permitted to exceed the normal limits, which is attached to the upload. Or something along those lines. There could also be more granular limits. For instance, local user group members could be permitted to grant tokens for their local area. Larger changes could require a national user group token. I would be in favor of applying granular limits on a location basis to protect from vandalism.

**Comment by Chaz6** [comment]
Another suggestion is for large changesets to be placed in an approval queue, which again could be done on a local or national user group basis depending on the size of the change.

**Comment by mmd-osm** [comment]
Approval queues have been suggested by a number of people in the past. I believe the concept would be extremely expensive to implement, it would cause massive disruption all over, hence it is simply not feasible. Besides, there's noone to review changes in time, and then those queued changed become stale and conflict with other changes that have been done in the meantime. Good luck resolving those.

**Comment by bhousel** [comment]
@Chaz6 These are actually very good suggestions, and very similar to what [Waze does](https://wazeopedia.waze.com/wiki/Benelux/Editor_level) with their editor levels (advanced editing features are locked until you attain a certain editor score).

It's not something we can easily add to OSM, given the project's inertia, but if I were starting over from scratch I'd probably implement something similar from the get-go.

**Comment by mmd-osm** [comment]
> It's not something we can easily add to OSM

@bhousel : you could still do this entirely in iD, and only enable some features after a user passes the respective parts of the tutorial, or by using some other metrics. After some time, you could encourage a user to learn something new, and then enable that feature. It's a bit of an adaptive tutorial mode.

**Comment by bhousel** [comment]
> @bhousel : you could still do this entirely in iD, and only enable some features after a user passes the respective parts of the tutorial, or by using some other metrics. After some time, you could encourage a user to learn something new, and then enable that feature. It's a bit of an adaptive tutorial mode.

If you want to talk to me about iD, please PM me.  I'm not really interested in starting a conversation about things we could add to iD, and this ticket isn't about iD anyway.

I was just responding to a person who I thought had some good suggestions and was getting some crappy feedback..  I'll unfollow this thread now.

**Comment by gravitystorm** [comment]
> Approval queues have been suggested by a number of people in the past. I believe the concept would be extremely expensive to implement, it would cause massive disruption all over, hence it is simply not feasible. Besides, there's noone to review changes in time, and then those queued changed become stale and conflict with other changes that have been done in the meantime. Good luck resolving those.

I completely disagree about review queues for changesets. We already have a need for send-and-poll for changeset uploads, [as I've described in a blog post](https://blog.gravitystorm.co.uk/2019/10/17/smoother-api-upgrades-for-openstreetmap/) and discussed with various editor developers. If that was implemented, then there is an opportunity for changesets to be reviewed before being applied. I would intend for that to be default-approve, immediately at first, and then with longer windows (e.g. 1 minute) as we develop processes and capacity to handle them. We could have longer timeouts based on changeset size, user status, all kinds of things.

But unfortunately this is all pie in the sky while we have so little development capacity available. So let's label this as "future" and return to more achievable goals.

We could do with having some default limits for all kinds of things - number of diary entries per day, number of gpx uploads, number of notes closed per day, number of changesets per day. It's not sustainable having all of these as "completely unlimited". We should add some safety rails.

This isn't a new concept, we already have it for messages per hour, for example.

**Comment by mmd-osm** [comment]
> We already have a need for send-and-poll for changeset uploads, as I've described in a blog post 

I think I disagree with this part of your blog. Uploading, storing and polling for updates would take more time than simply waiting for the result for a few seconds, now the changeset upload is no longer handled by Rails. This also works with compressed uploads on 2G mobile networks (I tested this part before).

Maybe I should have written that upload queues are not feasible _at this time_, as there's really lots of work involved to make that happen. And here I agree again with you, this is all pie in the sky at this time.

**Comment by woodpeck** [comment]
I agree with bhousel (minus the unnecessary negativity towards the pace of development and the i-am-publicly-unfollowing-this-now-because-you-are-not-worthy-of-my-time antics): This is not something we should be adding to editors because every editor will implement their own concept of what users to trust with what level of edits, and script writers will just circumvent any such limitations anyway. It is something we should have on a basic API level, just as we limit the maximum number of changes in a changeset.

And with "it" I mean a basic rate limiting for changes uploaded at first, where users only make e.g. 10k edits a day. Or make that 100k, and later reduce the limit as we find the time to code exceptions like mentioned above, where e.g. a well planned import process can receive an override token or whatever.

**Comment by mmd-osm** [comment]
Uhm, I thought Waze was locking down certain types for features, which I find very hard to do on the API level. I think we agree that only considering the number of edits to be a viable option in any case.

**Comment by simonpoole** [comment]
WRT editor integration, repeating what I pointed out on the relation size limit issue, and elsewhere, any kind of limit/block/whatever that potentially will cause the contributor to loose their work, needs to be detectable by the editing application in advance. This will allow a app to prevent the user in a suitable form from continuing to edit, or at least point out the potential frustration.

This ranges from maximum upload limits as originally discussed, over changeset review queues (which for example would need to block further contributions from editors in the same area), to any kind of data model limits (relation member size limits for example).

**Comment by richlv** [comment]
Unless we have time and resources to implement something fancy, may I suggest to temporarily shelve anything that takes more than 2 weeks real time to get on osm.org?

I love overengineering things, and https://xkcd.com/974/ .
But:
a) There's a high risk that this will be heatedly debated 2, 3, 5 years from now.
b) Something simple in place might prevent only a fraction of issues, but that's more than nothing.
c) Something simple is easier to extend [if following a sane design].

For example, let's start with a hardcoded limit, set in the DB.
Then it could be extended with per-user overrides, still directly in the DB. If somebody gets to it.
Then some admin interface could be added. If somebody gets to it.

**Comment by Lee-Carre** [comment]
> people — usually, but not always new signups — upload hundreds of thousands of objects to OSM before someone notices and tells them to stop.

So make the limits conditional upon account age, then.

This would also discourage the mischievous from opening new accounts in order to circumvent restrictions.

> we should really focus a bit more on UX and discuss ways on how to guide (new) users without patronizing them too much.

Definitely. See [openstreetmap/iD#8590 (comment)](/openstreetmap/iD/issues/8590#issuecomment-923436541) & subsequent discussion.

**Comment by richlv** [comment]
Just a small reminder that blindly limiting by account age would negatively affect proper imports that create dedicated accounts. Solutions to that discussed earlier.

**Comment by matkoniecz** [comment]
Given that proper import should document import (including account that will be used) and wait some time for feedback from community - proper imports will not be impacted.

**Comment by ZeLonewolf** [comment]
This issue was mentioned in the community forum:
https://community.openstreetmap.org/t/dwg-username-impersonation/101212

**Comment by scaidermern** [comment]
And also here: https://community.openstreetmap.org/t/how-about-limit-new-accounts/101656/

**Comment by matkoniecz** [comment]
Preamble: I am aware that repository maintainers are aware of facts stated in this comment. This comment is attempt to summarize what was posted so far and propose some specific limits. I made it with hope that it will be useful.

So we have few types of edits:
- regular low intensity edits, should not trigger rate limit
- DWG, can be exempt from rate limit
- power users, for example people reverting vandalism. Ideally they can be issued some flag but for now maybe just exempt from limits accounts with more than 2000 edits? (also, once changeset rate limit exists: maybe such accounts can be exempt from changeset comment limit or get it relaxed?)
- new users accounts that should be able to survive with small allowed edit rate
- bot accounts, especially new one can be tricky. Rate limits will hit them. But is it really so bad to ensure that new bot accounts are forced to go slowly?

For specific limits:
- DWG and accounts with more than 5000 nonepty changesets: no limits or something like 6000 changesets/h and 5M affected elements/h
- people with more than 2000 changesets: 600 changesets/h, 500k affected elements/h - first six additional unresolved reports filed about them reduces it, each by half (maybe count only one report from given account? Either way floor is needed here to avoid locking account by report spam. And reports from blocked accounts should be ignored here.)
- people with more than 100 changesets: 100 changesets/h, 50k affected elements/h limit (reports affect them in the same way)
- people with less than 100 changesets: 10 changesets/h, 5k affected elements/h limit

Costs here are additional complexity, also for editors that need to handle this (one it starts being developed it would make sense to ping at a least iD maintainer, likely also other developers). Mostly by providing new error message. The same goes for API libraries.

Is it viable for new legitimate user to edit more than 5000 elements and try to submit as their first edit and become stuck

Maybe it would be feasible to reduce limits by factor of 10 and count each tagless node as 1/100 of object? That would reduce vandal bit impact without risk of blocking new people mapping landuse.

Is it possible to assign greater penalty to empty edits? This is associated with buggy software and vandals. For example by counting opening changeset as 5, then reducing count by 4 once first data is send.

For future: rate limit exempt flag. Would it make sense to discuss with community/DWG/whoever relevant how it would be assigned? Or is it better to wait for at least initial implementation?

Would it make sense to include also IP-based rate limit, not only per account rate-limit?

Considered: button to temporarily bypass rate limits (complex, will help in small subset of cases), taking into account OSMF membership (not happy about this starting to give any privileges in mapping). Date of registration - reduce limits for new accounts (additional complexity, seems easy to work around it but may make sense).

**Comment by matkoniecz** [comment]
(if proposal above is too long/too complex: let me know and I will try to make more compact one keeping the most important parts)

**Comment by mmd-osm** [comment]
Initially, I thought this could be managed by the "changeset create" endpoint only, which is implemented both on CGImap and Rails (we're using the Rails version in production).

However, since we have apps like StreetComplete, we need to be careful to have a sane limit of currently opened changesets at any one time. If this limit is too low, it will break StreetComplete. Also, changesets might live up to 24 hours, if you keep on uploading small changes over a period of multiple hours, i.e. we need to be more precise what exactly "changesets/h" means. 

Without applying additional changes to CGImap, I don't see how we cold possibly limit the number of max. changes for each of these open changesets to less than 10k changes, or have a limit on the total number of changes per hour.

Bottom line, these changes are much more difficult to implement than it might appear.

**Comment by simonpoole** [comment]
IMHO the limit should be based on (total) number of objects changed not changesets. 

Naturally this doesn't solve the related changeset issues, but I don't think that the changeset is "unexpectedly" closed is actually one of them, as all editors already need to be able to deal with that in any case (even if they don't explicitly close changesets). 

What does need investigating (and this is something we can do now) is what is the most user friendly way to abort/refuse an upload in the current editor/API constellation. "user friendly" as in legit edits remain in the editor and are not lost.

**Comment by mmd-osm** [comment]
>  "user friendly" as in legit edits remain in the editor and are not lost.

Good question, to which I don't have an answer right now. It might be something based on HTTP 409 or HTTP 412, and  include an error messages stating the number of minutes the user has to wait before retrying. In any case, any editor currently applying some regex magic to extract meaningful data out of these error messages might face some issues here.

Let's not forget, that we need to cover the single object endpoints on Rails as well next to the diff upload (both Rails and CGImap).

**Comment by matkoniecz** [comment]
> Bottom line, these changes are much more difficult to implement than it might appear.

If I would consider them simple then I would submit PR rather than writing hopefully-useful-comments :)

> Overall, I'd also like to mention that I didn't see a thorough analysis of potential attack vectors, still we're already discussing some detailed solutions here, which feels to me much like an ad-hoc attempt at fixing things. That's not exactly the kind of working model I'd like to see for security topics.

Is it about some dedicated attacks like initiating multiple huge uploads at the same time, or opening multiple changesets and keeping them alive to create big stack and using them all at once?

Or about kind of attacks we want to stop (bulk vandalbot like latest attack, badly done import etc) 

Os is it about exploiting this method to cripple regular user who is not malicious (false reporting etc)?

Or about something else?

**Comment by mmd-osm** [comment]
> Or about something else?

Basically all of that, and much more. The idea is to make an informed decision by first assessing potential threats, and then working out potential ways on how to mitigate them. We should also keep in mind that a 100% perfect solution is completely unrealistic, and that every kind of limit might be impacting some legitimate users.

**Comment by matkoniecz** [comment]
Hmm. I worried that https://github.com/openstreetmap/openstreetmap-website/issues/2342#issuecomment-1694650004 may be overly long and unhelpful, it seems that problem may be that it is too short and incomplete in describing actors/scenarios. I can try to expand it then, if that would be considered as useful.

> 100% perfect solution is completely unrealistic

Oh, not expecting that at all.

**Comment by mmd-osm** [comment]
Yes, you could start working on these scenarios, and then discuss any specifics with the maintainers offline. We might also need some additional third party support for this task. I leave it to Rails port maintainers to decide what they deem necessary at this point in time.

**Comment by gravitystorm** [comment]
> if proposal above is too long/too complex:

My first comment here: It'll help developers if you swap all mentions of "DWG" with "moderator" - we don't have the concept of "DWG" in the codebase, and I need to keep mentally switching back and forth! I know there's not a 1:1 mapping between the two but that can be left for another time.

**Comment by HarelM** [comment]
Here's another use case where this solution could have a prevented a lot of grief:
https://community.openstreetmap.org/t/vandalism-and-blocks-in-israel/105176/8

Due to the current state in Israel a huge vandalism was caused by newly created users who remove tons of data from Tel Aviv and currented a lot of other places.
OSM volunteers are now trying to revert these vandalism acts, but it not easy not fast :-((

**Comment by zstadler** [comment]
One example that got through is a new user [vecago7283](https://www.openstreetmap.org/user/vecago7283) who was able to submit 99 changesets within 3 hours (between  2023-10-21T14:56:34Z and 2023-10-21T17:52:27Z). According to a simplistic Overpass Turbo query, these changesets included 216722 unique nodes, 426 unique ways and 288 unique relations. Many elements were modified multiple times, making the revert effort more complex than usual.

IMHO,
* Malicious edits need to be considered as an evolving permanent threat to the OSM data integrity.
* Just like viruses, it can be safely assumed that agenda-driven vandals will steadily improve they methods and learn the safeguards OSM has and will have put in place
* > Maybe it would be feasible to reduce limits by factor of 10 and count each tagless node as 1/100 of object? That would reduce vandal bit impact without risk of blocking new people mapping landuse.
   
   Modifying un-tagged nodes belonging to ways can create havoc in the map. I would not advise them to be counted with 1/100 multiplier.
   ![image](https://github.com/openstreetmap/openstreetmap-website/assets/1304610/63ee0381-1c17-4348-bcdc-c47c7a3c76d7)
* Perhaps a lower rate-limits should be made on deleting and modifying existing data and higher rate limits will apply to new data, which also happens to be easier to revert. Perhaps modifying an existing element would count as much as adding 10 new elements.
*  > Is it viable for new legitimate user to edit more than 5000 elements and try to submit as their first edit and become stuck

   Each element modification should be counted, not just the total number of elements modified.
   I also tend to disagree with this generosity towards new users.

**Comment by zstadler** [comment]
One example that got through is a new user [vecago7283](https://www.openstreetmap.org/user/vecago7283) who was able to submit 99 changesets within 3 hours (between  2023-10-21T14:56:34Z and 2023-10-21T17:52:27Z). According to a simplistic Overpass Turbo query, these changesets included 216722 unique nodes, 426 unique ways and 288 unique relations. Many elements were modified multiple times, making the revert effort more complex than usual.

IMHO,
* Malicious edits need to be considered as an evolving permanent threat to the OSM data integrity.
* Just like viruses, it can be safely assumed that agenda-driven vandals will steadily improve they methods and learn the safeguards OSM has and will have put in place
* > Maybe it would be feasible to reduce limits by factor of 10 and count each tagless node as 1/100 of object? That would reduce vandal bit impact without risk of blocking new people mapping landuse.
   
   Modifying un-tagged nodes belonging to ways can create havoc in the map. I would not advise them to be counted with 1/100 multiplier.
   ![image](https://github.com/openstreetmap/openstreetmap-website/assets/1304610/63ee0381-1c17-4348-bcdc-c47c7a3c76d7)
* Perhaps lower rate-limits should be made on deleting and modifying existing data and higher rate limits will apply to new data, which also happens to be easier to revert. Perhaps modifying an existing element would count as much as adding 10 new elements.
*  > Is it viable for new legitimate user to edit more than 5000 elements and try to submit as their first edit and become stuck

   Each element modification should be counted, not just the total number of elements modified.
   I also tend to disagree with this generosity towards new users.

**Comment by SomeoneElseOSM** [comment]
Just to give an idea of the scale of the problem, here's some vandalism from the last hour:

    changesets=> select id,created_at,num_changes from osm_changeset where user_name = 'bojicat389';
        id     |     created_at      | num_changes
    -----------+---------------------+-------------
     143253187 | 2023-10-28 10:23:25 |       10000
     143253195 | 2023-10-28 10:23:33 |       10000
     143253207 | 2023-10-28 10:23:42 |       10000
     143253219 | 2023-10-28 10:23:51 |       10000
     143253229 | 2023-10-28 10:23:59 |       10000
     143253238 | 2023-10-28 10:24:07 |       10000
     143253243 | 2023-10-28 10:24:14 |       10000
     143253252 | 2023-10-28 10:24:22 |       10000
     143253260 | 2023-10-28 10:24:30 |       10000
     143253271 | 2023-10-28 10:24:38 |       10000
     143253278 | 2023-10-28 10:24:46 |       10000
     143253284 | 2023-10-28 10:24:54 |       10000
     143253293 | 2023-10-28 10:25:02 |       10000
     143253302 | 2023-10-28 10:25:10 |       10000
     143253309 | 2023-10-28 10:25:18 |       10000
     143253317 | 2023-10-28 10:25:28 |        7833
    
    (snip, 39 changesets in total)

**Comment by tomhughes** [comment]
Look posting lots of lists of numbers doesn't help anybody.

Absolutely nobody is denying there is a problem.

Please just stop being unhelpful.

**Comment by ZeLonewolf** [comment]
> Look posting lots of lists of numbers doesn't help anybody.
> 
> Absolutely nobody is denying there is a problem.
> 
> Please just stop being unhelpful.

Numbers that characterize the problem are absolutely appropriate in an issue ticket discussing access limits.  That last comment was totally uncalled for and is part of the reason why it's so hard to attract contributors to the working group.

**Comment by zstadler** [comment]
The DWG members are working day and night to address organized vandalism in Israel, Ukraine, and probably elsewhere in the world. The ease in which these vandals create havoc in OSM is in an unbelievable contrast to the difficulty of fixing that. 

This issue is open for more than 3 years now.

@SomeoneElseOSM's comment is a wake-up call for all of us. It comes from the people who actually clean the sewage in order for the rest of us to have a clean map. It is the moral duty of the entire OSM community, and the people who can drive this issue in particular, to change gears and drive the issue to a resolution.

I'll leave you with a quote attributed to Voltaire and Montesquieu: [_"Perfect is the enemy of good"_](https://en.wikipedia.org/wiki/Perfect_is_the_enemy_of_good)

**Comment by matkoniecz** [comment]
@zstadler And this comment is just more elaborate "+1" comment and provides no useful technical info at all, just repeat of "it would be better to have that issue fixed and it is bad that it is not yet fixed".

**Comment by zstadler** [comment]
@matkoniecz ,
Fine. Please help me understand what can I, an potentially others, do in order to get this issue resolved?

Is there a list of gating issues?
Is there a decision process, or are we waiting for a concensus?

**Comment by HarelM** [comment]
If you read it thoroughly, it's not a simple +1, it suggests to approach this issue in an agile manor: start off with something simple, based on the given statistics in this thread, to help prevent the current on going problem, after that, iterate and find more sophisticated ways to improve the mechanism.
This can be easily compared to any cyber/virus attack: it's always easier to attack than to defend, and the more complicated the defence, so will the attack be, it's not a one-time thing we'll be able to solve with a single solution.
We need to start with something and improve it as we go along.
I'll be happy to pitch in, in any way, to push this forward.

**Comment by matkoniecz** [comment]
>  Fine. Please help me understand what can I, and potentially others, do in order to get this issue resolved?

Can you write C++ code? Are you willing to spend days/weeks/months on that? If no, then there is very limited amount of what such person can do. And if not, then writing about moral duty does not help much.

-----------------------------

As I understand (not an expert, I may be wrong) it requires relatively complex changes to https://github.com/openstreetmap/cgimap - written in C++ with all consequences of that.

For start, it would require significant time from someone who can write decent C++ code.

---------------

Further, maybe useless speculation based on:

> Just to add a bit of detail here, we already have rate limiting on the changeset upload call on cgimap, which is currently based on the number of bytes returnes by the call (like it is done for the /map call).
> Admittedly, looking at the number of bytes is not effective by any means to stop the users mentioned here (main reason being that the diffResult message usually comes with a rather tiny amount of bytes compared to a /map call). ;aybe we could evolve this approach by defining some more meaningful criteria.

maybe separate byte-counting could be done for performed edits? Maybe counting send data? Or by counting also send data, possibly with increased multiplier? With waiver for moderator accounts used by DWG?

**Comment by zstadler** [comment]
> For start, **it** would require significant time from someone who can write decent C++ code.

Let's assume I can write decent C++ code and I have the time for "it".

Let's try to nail down what an initial, minimal "it" is required to perform:
1. Is it a limit on:
   - The number of changesets?
   - The `changes_count` of the changesets?
   - The size in bytes of the change data (rate limit)?
   - or something else?
2. What are the initial limits?
3. How should they be increased?

I suggest defining the requirements in that order, one by one, rather than mix the discussion.

**Comment by tomhughes** [comment]
I've literally spent much of today analysing edits to try and figure out the best limits and start implementing them on the rails side - once we have that the cgimap side should be relatively easy as working out the limits is the hard part.

**Comment by zstadler** [comment]
@tomhughes,
Can you share what you figure out with respect the the 3 questions above?

> * Is it a limit on:
>   
>   * The number of changesets?
>   * The `changes_count` of the changesets?
>   * The size in bytes of the change data (rate limit)?
>   * or something else?
> 
> * What are the initial limits?
> 
> * How should they be increased?

**Comment by tomhughes** [comment]
It's on the number of changes - nothing else makes any sense.

The rest will mostly be configurable - it will be similar to the changeset comment limit which increase from an initial value to a maximum but I'm still working out what limits we might use and how the increase should happen.

**Comment by mmd-osm** [comment]
@tomhughes : @DavidKarlas has posted some good ideas as part of https://nightwatch-test.azurewebsites.net/RateLimit -

The number of {created,modified,deleted} {nodes,ways,relations} would be available in cgimap, at least for the current changeset. 24h sliding window might be a bit interesting to implement.

**Comment by zstadler** [comment]
@tomhughes wrote:
> It's on the number of changes - nothing else makes any sense.

I agree.

> I'm still working out what limits we might use and how the increase should happen

Thank you! As @HarelM [wrote above](https://github.com/openstreetmap/openstreetmap-website/issues/2342#issuecomment-1783882421)

> start off with something simple, based on the given statistics in this thread, to help prevent the current on going problem, after that, iterate and find more sophisticated ways to improve the mechanism.


@mmd-osm wrote:
> The number of {created,modified,deleted} {nodes,ways,relations} would be available in cgimap, at least for the current changeset. 24h sliding window might be a bit interesting to implement.

In the current flood of vandalism in Israel, there were very few (if any) created elements. They are also less complected to revert.
However, IMHO, considering the the different types of changes could be part of a future iteration, if it will be found applicable for the goals of that iteration.

**Comment by tomhughes** [comment]
I'm not trying to do anything as complicated as that algorithm, just a simple limit on the number of create/update/delete operations.

I also think that a 24 hour window is probably too long as it would need quite a high limit that people could then burst to very quickly so I was looking at a rolling one hour window with a limit ramping up from 1000 to 100000 or so over the first week when somebody starts at editing but possibly non-linearly and also with a reduction when users are reported as we did with changeset comments.

**Comment by DavidKarlas** [comment]
https://nightwatch-test.azurewebsites.net/RateLimit was not meant to go public yet, but it is OK if it is now... It was mostly for me to brainstorm and tweak limits to minimize false positives...
Reasoning for 24h limiting was mainly to give time to DWG to respond, but I didn't give more thought to window size.

I also think, looking at just changeset count is not enough, for example todays 3 vandals:
<img width="1377" alt="image" src="https://github.com/openstreetmap/openstreetmap-website/assets/774791/35c9dfcc-3df0-4b32-87f9-77ce9ee0b6a8">
None went past 20 changesets, but https://www.openstreetmap.org/user/Radwan%20Bohiba who is HOT mapper made 143 changesets in past 24h, and there are many that go past 100 changesets per day. Looking just at changeset count it will be very hard to minimize false positives...

Reason for tracking created/modified/deleted separately is mostly because modified limit needs to be much lower than created, because ease of revert but also legit mappers can create a lot when mapping buildings... Same goes for nodes/ways/relations...

Limits that I use today are maybe/probably order of magnitude too low, if I increased them by 10x, it would still catch most if not all vandals and almost no real mappers...

It also feels like mechanism for DWG to monitor which users were blocked in past 24h and give them ability to either ban or lift limitations for each account is almost must, otherwise we will fly blind and not know how many super active mappers are being discouraged from mapping...

**Comment by zstadler** [comment]
Another point to consider is the recovery speed. 

In a [recent revert](https://www.openstreetmap.org/api/0.6/changeset/143319333), 9002 changes were made in 86 minutes, or 105 changes/minute.
The relevant malicious account was able to modify about 250,000 elements in 80 minutes, or 3125 changes/minute
A 1000 changes/hour limit on new users is 17 changes /minute

That means that the revert rate would be more that 6 times faster than the malicious rate instead of being 30 times slower now!

**Comment by osmf-treasurer** [comment]
> I'm not trying to do anything as complicated as that algorithm, just a simple limit on the number of create/update/delete operations.
> 
> I also think that a 24 hour window is probably too long as it would need quite a high limit that people could then burst to very quickly so I was looking at a rolling one hour window with a limit ramping up from 1000 to 100000 or so over the first week when somebody starts at editing but possibly non-linearly and also with a reduction when users are reported as we did with changeset comments.

I'm currently preparing a talk for the SotM Europe about that and have screened the 600 or so most active users for that. In the last 11 years (since the license change)
* Only 8 probably benign users touched over 10000 objects in the first hour of editing activity
* Only 1 additional probably benign user touched over 40000 objects in the first 24 hours of editing activity

The metric actually used have been the number of created object versions, i.e. collating creations, changes, and deletions. Note that account creation on many cases have been hours to years ago, so the start of measurement is really the upload of the first changeset.

The 24 hour window stops users that first probe whether their setup works then start uploading massive changesets. If you decide to have a sliding window for all users and not just a limit for new users than the one-hour limit is probably good enough. I have not checked how many seasoned users have ever uploaded more than 10000 object versions in an hour.

Note that I do not count valid mechanical edits and approved imports towards benign users.

The Israel vandals are well over these values and would have been stopped in the third or fourth changeset if that limit were already in place. So would have been many Ukraine vandals.

**Comment by zstadler** [comment]
> Note that I do not count valid mechanical edits and approved imports towards benign users.

How were they identified as such? Manually? 
It will be interesting to see some statistics about them too.

**Comment by DavidKarlas** [comment]
I updated https://nightwatch-test.azurewebsites.net/RateLimit to use logic from https://github.com/openstreetmap/openstreetmap-website/pull/4319, at least try to(there is no API for first changeset time, hence had to calculate somehow, and maybe I made typo while converting algorithm...). It should be displaying all changesets that went past limit in last 6 days...

Anyway, looks like [Heinz_V](https://www.openstreetmap.org/user/Heinz_V) and few others went past 100k changes per hour upper limit.
Also new [HOT contributor](https://www.openstreetmap.org/user/TonyaRose) went past 1000 changes that new user get, users start getting more than 1000 changesets per hour after 16 hours and 48minutes from their first changeset.

**Comment by drolbr** [comment]
Comment https://github.com/openstreetmap/openstreetmap-website/issues/2342#issuecomment-1784533065 should have been from my developer account (this one) not osmf-treasurer. I'm sorry I've forgotten to check the account I'm logged in with.

**Comment by drolbr** [comment]
> > Note that I do not count valid mechanical edits and approved imports towards benign users.
> 
> How were they identified as such? Manually? It will be interesting to see some statistics about them too.

They were manually identified. I've looked into the user profile pages and classified all accounts that declare themselves to be bot accounts as such. Note that the two goods we are balancing here are
* the first contact for humans should not be bureaucratically repulsive by some arbitrary detection false positive
* malicious actors shall be stopped early enough such that reverts are still easy

In all cases where an account describes itself as being a bot account then a quota might impede the account but will not intimidate the human behind that account, because they then know OpenStreetMap and its rules already since some time. So limiting such an account is not a big issue. If we arrive at a mechanism with a pre-approved bot flag then we even can exempt such accounts from the limit.

**Comment by DavidKarlas** [comment]
I added information if account is Deleted or Blocked to https://nightwatch-test.azurewebsites.net/RateLimit so it should be much easier to see vandals vs. benign users cases...

**Comment by tomhughes** [comment]
> Anyway, looks like [Heinz_V](https://www.openstreetmap.org/user/Heinz_V) and few others went past 100k changes per hour upper limit.
> Also new [HOT contributor](https://www.openstreetmap.org/user/TonyaRose) went past 1000 changes that new user get, users start getting more than 1000 changesets per hour after 16 hours and 48minutes from their first changeset.

Do we really believe they are hand editing though?

I'm quite sure that Heinz_V isn't - it's clearly a bot because it uploads whole changesets full of nodes and only later uploads the ways and relations that use them.

**Comment by matkoniecz** [comment]
> it's clearly a bot because it uploads whole changesets full of nodes and only later uploads the ways and relations that use them.

Are they editing with JOSM? I am pretty sure that in case of massive changes JOSM splits it in multiple changesets this way, so this does not really indicate this.

Though data volume indicates bot/import anyway.

**Comment by zstadler** [comment]
https://github.com/openstreetmap/openstreetmap-website/pull/4319 says the limit is
> increasing as the square of time elapsed 

I suggest to use the accumulated number of changes rather than the elapsed time as the measure by which the rate is increased

I can see a malicious account creating a benign changeset, waiting a week doing nothing suspicious, and then receiving a decent rate. 
Alternatively, providing an initial 10000 benign edits, working for a period of at least 10 hours,  in order to get a rate increase is much harder to achieve. On the other hand, I do not believe that would be unacceptable for new inexperienced users.

**Comment by simonpoole** [comment]
> Are they editing with JOSM? I am pretty sure that in case of massive changes JOSM splits it in multiple changesets this way, so this does not really indicate this.

JOSM and fwiw Vespucci and I suspect any other editor that supports "auto-chunking" aka automatically splitting up changesets so that they are smaller than the maximum changeset limit will do this. While it would be better to do this with uploads that are referentially complete, I don't know of anybody that does that. See the longish discussion on making uploads idempotent somewhere on this repo for more information. BTW this is the mechanism that the vandals are using in Israel.

In any case this behaviour is one of the problems of introducing variable limits as proposed as the editor can't actually predict if the next upload will be successful or not, potentially leading to orphan nodes etc, because the parent elements can't be uploaded.

This is not a deal breaker, just something to keep in mind.

**Comment by mnalis** [comment]
> I updated https://nightwatch-test.azurewebsites.net/RateLimit to use logic from https://github.com/openstreetmap/openstreetmap-website/pull/4319, at least try to(there is no API for first changeset time, hence had to calculate somehow, and maybe I made typo while converting algorithm...). It should be displaying all changesets that went past limit in last 6 days...

@DavidKarlas quick look seems to show false positives on [StreetComplete](https://wiki.openstreetmap.org/wiki/StreetComplete) users, e.g. user [Lucidarius](https://hdyc.neis-one.org/?Lucidarius). Contributing factors:
- StreetComplete is popular editor which is made to be very simple as it is especially aimed on newbies, so it is quite likely it will be used on new accounts
- StreetComplete offers many (> 100) [quest types](https://wiki.openstreetmap.org/wiki/StreetComplete/Quests), and creates one-changeset-per-quest-type (e.g. one for `which way are lit`, one for `which surface is that`, one for `what are opening hours here` etc), which results in many changesets
- each changeset will contain many object modified (i.e. if user walked several km and answered `is this way lit` for each road segment of every road they passed, there could easily be hundreds of those in just one day. I myself often reach several hundreds, someone even over thousand objects edited in a day)
- often, users will upload the quests in batch only when they get home on wifi (as it saves battery on the field, reduces data usage and has other advantages related to undoing/modifying answers). (Or they might forget to do it that day and so do several other daily mapping sessions before finally doing the upload)

The combined result of those on the first look looks very much like vandalism would: few dozen changesets with hundreds of changes each (with no or very little additions/deletions), all uploaded by new user in several minutes; so it triggers with e.g. _"[Changeset 143292894](https://www.openstreetmap.org/changeset/143292894): Modified 798 ways in 1 hours, limit is 250."_

(more detailed processing reveals the StreetComplete changesets to be benign; but checking that is much more resource intensive -- i.e. requires looking what each object change did: StreetComplete will in vast majority of cases only add new tags like `lit=yes` or `lit/no`, and rarely do things modify `name` or delete tags or modify geometry or remove objects - which are likely more problematic)

**Comment by zstadler** [comment]
> quick look seems to show false positives on [StreetComplete](https://wiki.openstreetmap.org/wiki/StreetComplete) users, e.g. user [Lucidarius](https://hdyc.neis-one.org/?Lucidarius).

It seems to conflict with
> I updated https://nightwatch-test.azurewebsites.net/RateLimit to use logic from https://github.com/openstreetmap/openstreetmap-website/pull/4319

In the first day, that account has [created 847 changes](https://hdyc.neis-one.org/?Lucidarius), which it could have done on one hour, as described in https://github.com/openstreetmap/openstreetmap-website/pull/4319.
> defaults being a ramp up from 1000 changes per hour

**Comment by simonpoole** [comment]
General remark: there is no point in overthinking the criteria as they can all be trivially gamed, the best outcome is that there is an initial upper limit that keeps the volume of what needs to reverted more manageable than it is now, but it isn't going to stop a motivated vandal, just slow them down. 

> * StreetComplete is popular editor which is made to be very simple as it is especially aimed on newbies, so it is quite likely it will be used on new accounts

SC is used by ~5% of new sign ups with a peek above that now and then and newbies account for less than 30% of SC usage. It isn't going to be a large issue if the odd SC user would have to slow down a bit.

But that said, just applying the edit limits to anything outside of iD and the limited editing apps (MM, OM, SC, MC) only would likely reduce the false positive rate to near zero. Unluckily we don't really have a reliable way to identify the editing application and it is easy to masquerade as something else.

**Comment by mmd-osm** [comment]
> For start, it would require significant time from someone who can write decent C++ code.

No need to look into C++, we're doing it in a different way.

**Comment by HarelM** [comment]
While I understand the need to make sure new users are not blocked, I do want to convey the message of some of our veteran mappers in our OSM Israel group.
After spending a huge amount of time volunteering to map (years), the vandalism that happened in Israel and the time it took to revert it (which is still an on going effort) had made them question their motivation to continue mapping.
If you'll ask me, it's a lot more important to make sure their work and time spent was no in vain compared to newbies, and make sure they continue mapping, as they do it far better than new users.
While I'm trying to convince them that this can be fixed and that patience is needed here, this safeguard must be in place to prevent them from stopping to map.

As to how I see the rate limit (of course everything can be rigged, I don't doubt that):
I would say that the more experienced user you are the more edits you can make. The best measure I can think of regarding experience is the number of edits you see in your OSM profile page.
I usually use that to understand if a user is a new one or not, if the algorithm in the PR can use that, I think it is better than how much time passed since the user was created.

I would like to thank @tomhughes and @mmd-osm (and who ever else I probably missed) for pushing this forward! This is very much appreciated.

As a side note, what type of new users do a 10k change the first time they start to map? I honestly want to understand, not being cinic, I swear.
My edits are around 20-50 changes I think, and iD's save button turning yellow persuades me to save often...

**Comment by woodpeck** [comment]
> The best measure I can think of regarding experience is the number of edits you see in your OSM profile page.

Careful - that is the number of *changesets* that someone has, and you can easily have someone with hundreds of changesets with just one edit each - perhaps because they are using an app that does that, or they believe they need to "save" every change they make in iD - and on the other hand you can have someone with just a handful of changesets but a five-digit number of edits. The number is, sadly, relatively worthless.

**Comment by mmd-osm** [comment]
By the way, code is already deployed for testing here: https://tomh.apis.dev.openstreetmap.org/

You need to sign up there (your main osm.org account won't there), and then try uploading some data.

**Comment by mmd-osm** [comment]
Let's close this issue, rate limiting is in production now since Nov 15.
